Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -547,6 +548,81 @@ extension SwiftLanguageServer {
     // FIXME: cancellation
     _ = handle
   }
+
+  func foldingRange(_ req: Request<FoldingRangeRequest>) {
+
+    guard let snapshot = documentManager.latestSnapshot(req.params.textDocument.url) else {
+      log("failed to find snapshot for url \(req.params.textDocument.url)")
+      req.reply(nil)
+      return
+    }
+
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.editor_open
+    skreq[keys.name] = "IBSwiftAnnotationProvider:" + snapshot.document.url.path
+    skreq[keys.sourcetext] = snapshot.text
+    skreq[keys.syntactic_only] = 1
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else { return }
+      guard let dict = result.success else {
+        req.reply(.failure(result.failure!))
+        return
+      }
+
+      guard let syntaxMap: SKResponseArray = dict[self.keys.syntaxmap],
+            let substructure: SKResponseArray = dict[self.keys.substructure] else {
+        return req.reply([])
+      }
+
+      var ranges: [FoldingRange] = []
+
+      syntaxMap.forEach { _, value in
+        if let kind: sourcekitd_uid_t = value[self.keys.kind],
+           kind == self.values.syntaxtype_comment,
+           let offset: Int = value[self.keys.offset],
+           let start: Position = snapshot.positionOf(utf8Offset: offset),
+           let length: Int = value[self.keys.length],
+           // SourceKit marks the end of a comment as the first non-comment character
+           // after it, so we subtract one to get the real comment range.
+           let end: Position = snapshot.positionOf(utf8Offset: offset + length - 1) {
+          let range = FoldingRange(startLine: start.line,
+                                   startCharacter: offset,
+                                   endLine: end.line,
+                                   endCharacter: offset + length - 1,
+                                   kind: .comment)
+          ranges.append(range)
+        }
+        return true
+      }
+
+      var structureStack: [SKResponseArray] = [substructure]
+      while let substructure = structureStack.popLast() {
+        substructure.forEach { _, value in
+          if let offset: Int = value[self.keys.bodyoffset],
+             let length: Int = value[self.keys.bodylength],
+             length > 0,
+             let start: Position = snapshot.positionOf(utf8Offset: offset),
+             let end: Position = snapshot.positionOf(utf8Offset: offset + length) {
+            let range = FoldingRange(startLine: start.line,
+                                     startCharacter: offset,
```

Comment:
I'm pretty sure the start/index "character" are columns, not absolute positions.  Otherwise the comment in the spec about defaulting to the length of the line make no sense (also the line+character terminology matches what's in "Position", which is definitely a column index).

So these should be start.utf16index and end.utf16index I think.

Could this be related to what you mentioned in the description about how the line folding didn't match vscode's builtin folding?

Reply:
That makes a lot of sense! I got confused by the name they used in the protocol.
It didn't changed that behaviour though. I _assume_ they're ignoring these properties completely in favor of folding the entire line, because if you have something like this:
```
func foo() {
    func bar() {

    } //comment
}
```
and subtract one from the end offset (so it ends before the closing brace), `foo` will fold like the builtin version, but `bar` loses the brace and the comment. That would explain why things worked even though the characters were wrong.